src/main.rs

```diff
@@ -180,14 +185,143 @@ fn build_memory_map<T: GuestStagePagingMode>(fdt: &Fdt) -> MemMapResult<HwMemMap
+        | HwMemRegionType::Reserved(HwReservedMemType::HypervisorPtes)
+        | HwMemRegionType::Reserved(HwReservedMemType::PageMap) => {
+            Some((r.base(), r.size(), PteLeafPerms::RW))
+        }
+        HwMemRegionType::Mmio(_) => Some((r.base(), r.size(), PteLeafPerms::RW)),
+    }
+}
+
+// Adds an identity mapping to the given Sv48 table for the specified address range.
+fn hyp_map_region(
+    sv48: &FirstStagePageTable<Sv48>,
+    base: PageAddr<SupervisorPhys>,
+    size: u64,
+    // TODO - page permissions.
+    _perms: PteLeafPerms,
+    get_pte_page: &mut dyn FnMut() -> Option<Page<InternalClean>>,
+) {
+    let region_page_count = PageSize::Size4k.round_up(size) / PageSize::Size4k as u64;
+    // Pass through mappins, vaddr=paddr.
+    let vaddr = PageAddr::new(RawAddr::supervisor_virt(base.bits())).unwrap();
+    // Add mapping for this region to the page table
+    let mapper = sv48
+        .map_range(vaddr, PageSize::Size4k, region_page_count, get_pte_page)
+        .unwrap();
+    for (virt, phys) in vaddr
+        .iter_from()
+        .zip(base.iter_from())
+        .take(region_page_count as usize)
+    {
+        // Safe as we will create exactly one mapping to each page and will switch to
+        // using that mapping exclusively.
+        unsafe {
+            mapper.map_4k_addr(virt, phys).unwrap();
+        }
+    }
+}
+
+// Creates the Sv48 page table based on the accessible regions of memory in the provided memory
+// map.
+fn setup_hyp_paging(mem_map: &mut HwMemMap) {
+    let num_pte_pages = pte_page_count(mem_map);
+    let pte_base = find_available_region(mem_map, num_pte_pages * PageSize::Size4k as u64)
+        .expect("Not enough free memory for hypervisor Sv48 page table");
+    let mut pte_pages = mem_map
+        .reserve_take_pages(
+            HwReservedMemType::HypervisorPtes,
+            SupervisorPageAddr::new(RawAddr::from(pte_base)).unwrap(),
+            PageSize::Size4k,
+            num_pte_pages,
+        )
+        .unwrap()
+        .clean()
+        .into_iter();
+    // Create empty sv48 page table
+    let root_page = pte_pages.next().unwrap();
+    let sv48: FirstStagePageTable<Sv48> =
+        FirstStagePageTable::new(root_page.into()).expect("creating sv48");
+
+    // Map all the regions in the memory map that the hypervisor could need.
+    for (base, size, perms) in mem_map.regions().filter_map(hyp_map_params) {
+        hyp_map_region(&sv48, base, size, perms, &mut || pte_pages.next());
+    }
+
+    // TODO- discover these from DT and un-hard code?
+    map_fixed_device(UART_BASE, &sv48, &mut || pte_pages.next());
+    // TODO - reset device is hard codec in vm.rs
+    map_fixed_device(0x10_0000, &sv48, &mut || pte_pages.next());
+
+    // Install the page table in satp
+    let mut satp = LocalRegisterCopy::<u64, satp::Register>::new(0);
+    satp.set_from(&sv48, 0);
+    // Store SATP value for other CPUS. then load from global in start seconday
```

Comment:
"Store the SATP value for other CPUs. They load from the global in start_secondary"

Reply:
better, changed.